File: app/src/main/java/com/unsplash/client/network/response/ApiResponse.java
package com.unsplash.client.network.response;

import androidx.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import rx.Observable;

public class ApiResponse<T> {

    @SerializedName("error")
    private ApiError error;

    @Nullable
    @SerializedName("results")
    private T data;

    @Nullable
    public T getData() {
        return data;
    }

    private boolean isSuccess() {
        return data != null;
    }


    public Observable<ApiResponse<T>> handleError() {
        if (isSuccess()) {
            return Observable.just(this);
        } else {
            return Observable.error(error);
        }
    }
}

File: app/src/main/java/com/unsplash/client/network/response/ApiResponse.kt
package com.unsplash.client.network.response

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

import rx.Observable

class ApiResponse<T> {

    @SerializedName("error")
    @Expose
    var error: ApiError? = null

    @SerializedName("results")
    @Expose
    var data: T? = null

    private val isSuccess: Boolean
        get() = data != null


    fun handleError(): Observable<ApiResponse<T>> {
        return if (isSuccess) {
            Observable.just(this)
        } else {
            Observable.error(error)
        }
    }
}
